### PR TITLE
Block Processing Validity Conditions

### PR DESCRIPTION
**Pull request by rauljordan**
Hi all,

This PR encompasses a small part of #256, namely, the processing of incoming beacon blocks into a node. There are a few validity conditions to consider, including timestamp validity of the block, validity of a mainchain reference, and the validity of a previous hash processed by the beacon node.

This PR creates a `CanProcessBlock` function returning a boolean to be used in the node's p2p event loop.  

### Requirements for Merge

- [x] Tests for CanProcessBlock

### CODE REVIEW

File: beacon-chain/types/block.go [hunk 1]
```diff
@@ -1,6 +1,27 @@
...
 
-import "github.com/ethereum/go-ethereum/common"
+import (
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// Block defines a beacon chain core primitive.
+type Block struct {
```
Comment (terenc3t) [inline beacon-chain/types/block.go]: In the ECDC video, Vitalik and Justin mentioned there's no `header` and `body` for a beacon chain block. The whole thing is just a block with fields

File: beacon-chain/blockchain/core_test.go [hunk 2]
```diff
@@ -122,22 +135,26 @@ func TestMutateCrystallizedState(t *testing.T) {
...
+		t.Fatalf("Unable to setup db: %v", err)
 	}
 	db.Start()
 	beaconChain, err := NewBeaconChain(db.DB())
 	if err != nil {
-		t.Fatalf("unable to setup beacon chain: %v", err)
+		t.Fatalf("Unable to setup beacon chain: %v", err)
+	}
+
+	priv, err := crypto.GenerateKey()
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply beacon-chain/blockchain/core_test.go]: I agree, good to note

File: beacon-chain/blockchain/core.go [hunk 3]
```diff
@@ -81,6 +92,24 @@ func (b *BeaconChain) MutateCrystallizedState(crystallizedState *types.Crystalli
 	return b.persist()
 }
 
+// CanProcessBlock decides if an incoming p2p block can be processed into the chain's block trie.
+func (b *BeaconChain) CanProcessBlock(fetcher powchain.POWBlockFetcher, block *types.Block) (bool, error) {
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply beacon-chain/blockchain/core.go]: Yep but there won't be anything to cancel or to use. The ctx argument will be unused so there is no point to passing it in.

[the remark this answers is not in the document]
Comment (rawfalafel) [reply beacon-chain/blockchain/core.go]: Is `context.Background` used by `fetcher.BlockByHash`? After #280, if the process receives a `SIGTERM` while inside `fetcher.BlockByHash`, the function should immediately return an error and allow the goroutine to exit cleanly. That requires the context to be passed down.

[the remark this answers is not in the document]
Comment (rauljordan) [reply beacon-chain/blockchain/core.go]: Sorry I missed this, @rawfalafel, I'll add it in my new PR on block processing.

**Pull request merged**